### ISSUE TITLE
Layer level security for portal / agol

### ISSUE THREAD

**Issue by keithfraley** [koopjs/koop#315]
First, really excited about the potential for koop to held bridge the gap between several systems we maintain.  

I have talked to Danny about leveraging Passport JS to enable layer level authentication for Koop services in Portal or AGOL.  Ideally, we would use the interface for embedding username and passwords in items that already exists today for Portal and AGOL.  

Danny mentioned that this is possible and I want to start this dialog to get this ball rolling :).

**Comment by fsxdrone** [comment]
In continuation of what Keith was saying... i was trying to implement **Bearer tokens from passportjs** and to get callback in getData working...ran into an error **{"error":"features.forEach is not a function"}**. I was trying to layer level json info to display the security info like this
{
  "secured": true,
  "httpStatusCode": 499,
  "httpStatusMessage": "Token Required",
  "httpResponse": "",
  "url": "xxxx/FeatureServer/0?f=json"
}

**Comment by keithfraley** [comment]
Any idea when timeframe is on this enhancement?

**Comment by rgwozdz** [comment]
@keithfraley hoping to get started on this in earnest this week. @fsxdrone - anything you can share about what you have tried so far would be interesting to see.  What does your passportjs implementation look like?

**Comment by rgwozdz** [comment]
Actually pretty close.  I would guess something early next week.

**Comment by rgwozdz** [comment]
Token based service security has now been implemented.  Koop 3.7.0 includes support for a new plugin-type `auth` and we have released an example auth-plugin at [koop-auth-direct-file](https://github.com/koopjs/koop-auth-direct-file).  This example plugin, along with updates to koop-output-geoservices and FeatureServer,  provides support for securing services in a manner that works with ArcGIS Online and ArcGIS Portal.
